hail/src/main/scala/is/hail/cxx/StagedBaseStructBuilder.scala

```diff
@@ -3,44 +3,64 @@ package is.hail.cxx
 import is.hail.expr.types.physical.PBaseStruct
 import is.hail.utils.ArrayBuilder
 
-class StagedBaseStructBuilder(fb: FunctionBuilder, pStruct: PBaseStruct) {
-  val ab = new ArrayBuilder[Code]
-  val s = Variable("s", "char *", s"${ fb.getArg(1) }->allocate(${ pStruct.alignment }, ${ pStruct.byteSize })")
-  ab += s.define
-  ab += s"memset($s, 0, ${ pStruct.nMissingBytes });"
+class StagedBaseStructBuilder(st: Code, pStruct: PBaseStruct, off: Expression) {
 
-  private var i = 0
+  def setAllMissing(): Code =
+    s"memset($off, 0xff, ${ pStruct.nMissingBytes });"
 
-  def add(t: EmitTriplet) {
-    val f = pStruct.fields(i)
-    ab +=
-      s"""
-         |${ t.setup }
-         |if (${ t.m })
-         |  ${
-        if (pStruct.fieldRequired(i))
-          "abort();"
-        else
-          s"store_bit($s, ${ pStruct.missingIdx(i) }, 1);"
-      }
-         |else
-         |  ${
-        storeIRIntermediate(f.typ,
-          s"($s) + (${ pStruct.byteOffsets(i) })",
-          t.v)
-      };
-         |""".stripMargin
-    i += 1
+  def clearAllMissing(): Code =
+    s"memset($off, 0, ${ pStruct.nMissingBytes });"
+
+  def setMissing(idx: Int): Code = {
+    if (pStruct.fieldRequired(idx))
+      s"""NATIVE_ERROR($st, 1010, "Required field cannot be missing.");"""
```

Comment:
The NATIVE_ERROR macro just sets up the NativeStatus, it doesn't have control flow.  You need to return (nullptr, probably) here.  Eventually we'll want a standard idiom where the entrypoint does a try/catch and sets st, and this just throws an exception.